RecoEgamma/EgammaHLTProducers/plugins/EgammaHLTPixelMatchVarProducer.cc

```diff
@@ -41,6 +41,8 @@ namespace {
       int layerOffset = 3;
       if (seed.subDet(hitNr) == PixelSubdetector::PixelEndcap)
         layerOffset += 4;
+      //layerOrDiskNr(hitNr) return numeric_limits<T>::max() only if hitNr>=seed.hitInfo().size()
+      [[clang::suppress]]
```

Comment:
clang anaylyzer complain that `layerOrDiskNr` can return `numeric_limits<T>::max()` if `hitNr>=seed.hitInfo().size()` but are are in a loop where `hitNr<seed.hitInfo().size()`. So this seems like a false positive warning

Reply:
I suppose this is supposed to partly fix https://github.com/cms-sw/cmssw/issues/46037

I am not a fan of these attribute specifiers. Isn't there a way to suppress the warning upstream or at least specify which kind of warning it's going to be suppressed?

Reply:
maybe this check can be rethought?
https://github.com/cms-sw/cmssw/blob/ea0fd48a9eb0dd1f7cf7d026fd28cda4140ff206/DataFormats/EgammaReco/interface/ElectronSeed.h#L143

[is it clear why silently returning garbage is preferred ? (or do we just want to avoid throwing here?)]

Reply:
@Sam-Harper do you have any insight on this? You introduced this function back in https://github.com/cms-sw/cmssw/pull/18541 (albeit the default return was changed from ` std::numeric_limits<T>::infinity()` to ` std::numeric_limits<T>::max()` in https://github.com/cms-sw/cmssw/pull/45488)

Reply:
> I suppose this is supposed to partly fix #46037
> 
> I am not a fan of these attribute specifiers. Isn't there a way to suppress the warning upstream or at least specify which kind of warning it's going to be suppressed?

@mmusich , unfortunately the only way to suppress such warnings are either using `[[clang::suppress]]` or using `#ifndef __clang_analyzer__  ..... #endif` . I can update the comment that it suppress large BitWiseShift warnings.

about mixing it upstream, I guess we can add `assert(hitNr < hitInfo_.size())` in the https://github.com/cms-sw/cmssw/blob/ea0fd48a9eb0dd1f7cf7d026fd28cda4140ff206/DataFormats/EgammaReco/interface/ElectronSeed.h#L142-L144 e.g.
```
    T getVal(unsigned int hitNr, T PMVars::*val) const {
      assert(hitNr < hitInfo_.size());
      return hitInfo_[hitNr].*val;
    }
```
this should fix the clang warning but I am not sure this that is right implementation

Reply:
or we can use `.at()` instead of `operator[]` - the difference is that `.at()` does range checking and raises `std::out_of_range`. No idea though if it will silence analyzer warnings.

Reply:
> or we can use `.at()` instead of `operator[]`

let's not slow down the HLT please.

Reply:
one other way to fix this is using something like
```
     T getVal(unsigned int hitNr, T PMVars::*val) const {
      return hitNr < hitInfo_.size() ? hitInfo_[hitNr].*val : 
        #ifndef __clang_analyzer__
          std::numeric_limits<T>::max();
        #else
          0;
        #endif
     }
```

but this will suppress this warning for code where we do actually call this function with `hitNr >= hitInfo_.size()` . I think best approach is to suppress it as suggested by this PR

Reply:
>  right - that change was also to avoid a clang warning for undefined behavior it seems:)

It looks like https://github.com/cms-sw/cmssw/pull/45488 ultimately caused https://github.com/cms-sw/cmssw/issues/45658. Perhaps indeed that code needs to be looked at more carefully.